File: apps/api/models/updates.py
"""Update-check wire models (ADR-044).

Mirrors ``doc_assistant.update_check.UpdateStatus`` plus the one user-settable field. The
TypeScript contract lives at ``apps/desktop/src/lib/core/types/updates.ts`` and changes in the
same commit as this file (``apps/api/CLAUDE.md``).
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel


class UpdateStatusOut(BaseModel):
    """What the app knows about whether a newer release exists.

    ``state`` has three values, not two. ``unknown`` means "never checked" or "the check failed" —
    a distinction the user does not need, since the action is the same. What it must never mean is
    ``current``: reporting "up to date" because the network was down is the failure ADR-044 exists
    to prevent, so ``reason`` carries the plain-words explanation instead.
    """

    state: Literal["current", "update_available", "unknown"]
    current_version: str
    latest_version: str | None = None
    release_url: str
    checked_at: str | None = None
    reason: str | None = None
    # Whether the *automatic* daily check is on. A manual check runs regardless (ADR-044).
    auto_check_enabled: bool


class UpdateSettingsUpdate(BaseModel):
    """Turn the automatic daily check on or off. The only writable field in this domain —
    the repository checked against is deliberately not configurable (ADR-044: pointing an update
    banner at an arbitrary host is a way to get someone to install something they did not choose).
    """

    auto_check_enabled: bool

File: apps/desktop/src/lib/core/types/updates.ts
// TypeScript mirror of the update-check payloads (apps/api/models/updates.py, ADR-044).
// Keep in sync with the pydantic models — this is the wire contract; a change to the model and
// a change here belong in the same commit (apps/desktop/CLAUDE.md).

/**
 * Three states, not two. `unknown` covers "never checked" and "the check failed" — the same
 * thing to a user deciding whether to go look. What it must never collapse into is `current`:
 * saying "up to date" because the network was down is the failure ADR-044 exists to prevent,
 * so the UI reads `reason` in that case instead of inventing reassurance.
 */
export type UpdateState = 'current' | 'update_available' | 'unknown'

export interface UpdateStatus {
  state: UpdateState
  current_version: string
  /** The newest published release, when one was seen. Null whenever `state` is 'unknown'. */
  latest_version: string | null
  /** Where "Get the update" goes. The app never downloads or installs anything (ADR-044). */
  release_url: string
  /** ISO stamp of the last completed check — how old this answer is, not when it was rendered. */
  checked_at: string | null
  /** Plain-words explanation, set only for 'unknown'. Safe to show verbatim. */
  reason: string | null
  /** Whether the *automatic* daily check is on. A manual check runs regardless. */
  auto_check_enabled: boolean
}
